### PR TITLE
Feature: whisper ver5 support

### LINKED ISSUE #65: Integrate with whisper ver. 5
**Objective:**
- once Whisper ver. 5 stabilizes, we need to update bindings code to use it, instead of ver. 2

**Notes:**
- ~~particularly watch for this PR to merge https://github.com/ethereum/go-ethereum/pull/3224~~

### PR DESCRIPTION
**Pull request by farazdagi**
- merge after https://github.com/status-im/status-go/pull/127
- closes #65 
- test suite has been extended:
  - all tests are done in jailed context (the same context RN app uses)
  - cases tested:
    - K1 -> K2, signed+encrypted message, from us
    - K2 -> K1, singed+encrypted message, to us
    - K1 -> K1, signed+encrypted message, to ourselves
    - K1 -> "", signed broadcast
    - "" -> "", anonymous broadcast
    - "" -> K1, encrypted anon broadcast
- breaking test added:
  - on multiple concurrent addition of filters, Whisper2 breaks #130 
  - separate test has been added, to make sure that same behavior doesn't exist on Whisper5
- web3js extended:
  - symmetric key management methods exposed:
    - `shh.hasSymKey(keyname string)`
    - `shh.generateSymKey(keyname string)`
    - `shh.addSymKey(keyname string, keydata string)`
    - `shh.deleteSymKey(keyname string)`
  - fixed `shh.post()` (`keyname` parameters added)
- Whisper2 -> Whisper5 migration notes:
  - in shh/2 we can sign (if we provide `from`) and encrypt messages (if we provide `to`), and if we omit `to` and/or `from` we could have unencrypted and unsigned messages. It is no more the case with shh/5 - **all** messages are encrypted now, this means that when **broadcasting** (i.e. , not specifying recipient using empty `to`), you need to specify `keyname` - the name of symmetric key to be used to encrypt message with. 
  - basically, all you need to update is:
    - add symmetric key using either `generateSymKey()` or `addSymKey()`. The former can be used to generate a session key (as @jarradh pointed out, session keys will allow us to have a forward secrecy, when leaked secret doesn't compromise previous communication), which you than need to send using asymmetric encryption
    - once symmetric key is present, you need to specify its name via `keyname` parameter (in both filter and message objects)
    - when it comes to asymmetric encryption that nothing changes - you provide `to` and `from` (former is used for signing, and the later is used for encryption, to be decrypted by recipient)

### CODE REVIEW

**Comment by farazdagi** [comment]
**Example using updated Web3.js API:**

```html

<html lang="en">
<head>
    <title>web3.js sample</title>
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <script
            src="https://code.jquery.com/jquery-3.1.1.js"
            integrity="sha256-16cdPddA6VdVInumRGo6IbivbERE8p7CQR3HzTBuELA="
            crossorigin="anonymous"></script>
    <script type="text/javascript" src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="./bignumber.js"></script>
    <script type="text/javascript" src="./web3.js"></script>

    <script type="text/javascript">
        var web3 = new Web3();
        var shh = web3.shh;
        web3.setProvider(new web3.providers.HttpProvider('http://localhost:8645'));

        var identity = '0x04eedbaafd6adf4a9233a13e7b1c3c14461fffeba2e9054b8d456ce5f6ebeafadcbf3dce3716253fbc391277fa5a086b60b283daf61fb5b1f26895f456c2f31ae3';
        if (!web3.shh.hasIdentity(identity)) {
            throw 'idenitity "0x04eedbaafd6adf4a9233a13e7b1c3c14461fffeba2e9054b8d456ce5f6ebeafadcbf3dce3716253fbc391277fa5a086b60b283daf61fb5b1f26895f456c2f31ae3" not found in whisper';
        }

        var topic = 'example3';
        var payload = 'test message 3 (K1 -> "", signed broadcast)';

        // generate symmetric key (if doesn't already exist)
        if (!shh.hasSymKey(topic)) {
            shh.addSymKey(topic, "0xdeadbeef"); // alternatively: shh.generateSymKey("example3");
                                                // to delete key, rely on: shh.deleteSymKey(topic);
        }

        // start watching for messages
        var filter = shh.filter({
            from: identity,
            topics: [web3.fromAscii(topic)],
            keyname: topic // you can use some other name for key too
        });
        filter.watch(function(error, result){
            if (!error) {
                console.log("Message received1: ", result);
            }
        }, function (error) {
            console.log("error callback")
        });


        setTimeout(function () {
            var message = {
                from: identity,
                topics: [web3.fromAscii(topic)],
                payload: payload,
                ttl: 20,
                keyname: topic
            };
            var err = shh.post(message)
            if (err !== null) {
                console.log("message NOT sent")
            } else {
                console.log("message sent OK")
            }
        }, 3000)


        $(document).ready(function () {
        });

    </script>

</head>

<body>
<!-- Static navbar -->
<nav class="navbar navbar-default navbar-static-top">
    <div class="container">
    </div>
</nav>

<div class="container">
    <div class="jumbotron"></div>
</div>
</body>
</html>
```

**Comment by farazdagi** [comment]
- @rasom I've deployed stand alone Whisper nodes on our cluster, which are used as hubs to deliver messages through (currently there's hardly many shh/5 nodes out there)
- I've added those 10 stand-alone nodes to our bootstrap list i.e. `status-go` knows those peers
- **CRUCIAL:** Please when testing, test that messages are really delivered, not only within one device, but across two devices i.e. that our cluster is used for delivery
- I did actually test that (but please, still double test before merging), using:

```html

<html lang="en">
<head>
    <title>Whisper Message Delivery Test</title>
    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css">
    <script
            src="https://code.jquery.com/jquery-3.1.1.js"
            integrity="sha256-16cdPddA6VdVInumRGo6IbivbERE8p7CQR3HzTBuELA="
            crossorigin="anonymous"></script>
    <script type="text/javascript" src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="./bignumber.js"></script>
    <script type="text/javascript" src="./web3.js"></script>

    <script type="text/javascript">
        var identity = '0x04eedbaafd6adf4a9233a13e7b1c3c14461fffeba2e9054b8d456ce5f6ebeafadcbf3dce3716253fbc391277fa5a086b60b283daf61fb5b1f26895f456c2f31ae3';
        var topic = 'example3';
        var payload = 'test message (sent via one node, received via another), message id: ';

        function senderLoop() {
            var web3 = new Web3();
            web3.setProvider(new web3.providers.HttpProvider('http://localhost:8645'));

            // check identity
            if (!web3.shh.hasIdentity(identity)) {
                throw 'idenitity "' + identity + '" not found in whisper';
            }

            // generate symmetric key (if doesn't already exist)
            if (!web3.shh.hasSymKey(topic)) {
                web3.shh.addSymKey(topic, "0xdeadbeef");
            }

            // start sending messages
            var cnt = 0;
            var timer = setInterval(function () {
                cnt++;

                var message = {
                    from: identity,
                    topics: [web3.fromAscii(topic)],
                    payload: payload + cnt,
                    ttl: 20,
                    keyname: topic
                };
                var err = web3.shh.post(message)
                if (err !== null) {
                    console.log("message NOT sent")
                } else {
                    console.log("message sent OK")
                }

                // max messages limit
                if (cnt === 5) {
                    clearInterval(timer);
                }
            }, 5000);
        }

        function watcherLoop() {
            var web3 = new Web3();
            web3.setProvider(new web3.providers.HttpProvider('http://localhost:8745'));

            // check identity
            if (!web3.shh.hasIdentity(identity)) {
                throw 'idenitity "' + identity + '" not found in whisper';
            }

            // generate symmetric key (if doesn't already exist)
            if (!web3.shh.hasSymKey(topic)) {
                web3.shh.addSymKey(topic, "0xdeadbeef");
            }

            // start watching for messages
            var filter = web3.shh.filter({
                from: identity,
                topics: [web3.fromAscii(topic)],
                keyname: topic // you can use some other name for key too
            });
            filter.watch(function(error, result){
                console.log(error)
                if (!error) {
                    console.log("Message received: ", result);
                }
            });
        }

        $(document).ready(function () {
            senderLoop()
            watcherLoop()
        });

    </script>

</head>

<body>
<!-- Static navbar -->
<nav class="navbar navbar-default navbar-static-top">
    <div class="container">
    </div>
</nav>

<div class="container">
    <div class="jumbotron">

    </div>
</div>


</body>

</html>

```
